packages/eslint-plugin/tests/rules/require-types-exports.test.ts

```diff
@@ -0,0 +1,2007 @@
+      export function f(): A {}
+    `,
+
+    `
+      export type A = number;
+      export const f = (): A => {};
+    `,
+
+    `
+      export type A = number;
+      export type B = string;
+      export function f(): A | B {}
+    `,
+
+    `
+      export type A = number;
+      export type B = string;
+      export const f = (): A | B => {};
+    `,
+
+    `
+      export type A = number;
+      export type B = string;
+      export function f(): A & B {}
+    `,
+
+    `
+      export type A = number;
+      export type B = string;
+      export const f = (): A & B => {};
+    `,
+
+    `
+      export type A = number;
+      export type B = string;
+      export function f(): [A, B] {}
+    `,
+
+    `
+      export type A = number;
+      export type B = string;
+      export const f = (): [A, B] => {};
+    `,
+
+    `
+      export type A = number;
+      export type B = string;
+      export function f(): { a: A; b: B } {}
+    `,
+
+    `
+      export type A = number;
+      export type B = string;
+      export const f = (): { a: A; b: B } => {};
+    `,
+
+    `
+      import { testFunction, type Arg } from './module';
+
+      export function f(a: Arg): void {}
+    `,
+
+    `
+      import { Arg } from './types';
+
+      export function f(a: Arg): void {}
+    `,
+
+    `
+      import type { Arg } from './types';
+
+      export function f(a: Arg): void {}
+    `,
+
+    `
+      import type { ImportedArg as Arg } from './types';
+
+      export function f(a: Arg): void {}
+    `,
+
+    `
+      import type { Arg } from './types';
+
+      export function f<T extends Arg>(a: T): void {}
+    `,
+
+    `
+      export type R = number;
+
+      export function f() {
+        const value: { num: R } = {
+          num: 1,
+        };
+
+        return value;
+      }
+    `,
+
+    `
+      import type { A } from './types';
+
+      export type T1 = number;
+
+      export interface T2 {
+        key: number;
+      }
+
+      export const value: { a: { b: { c: T1 } } } | [string, T2 | A] = {
+        a: {
+          b: {
+            c: 1,
+          },
+        },
+      };
+    `,
+
+    `
+      import type { A } from './types';
+
+      export type T1 = number;
+
+      export interface T2 {
+        key: number;
+      }
+
+      const value: { a: { b: { c: T1 } } } | [string, T2 | A] = {
+        a: {
+          b: {
+            c: 1,
+          },
+        },
+      };
+
+      export default value;
+    `,
+
+    `
+      export enum Fruit {
+        Apple,
+        Banana,
+        Cherry,
+      }
+
+      export function f(a: Fruit): void {}
+    `,
+
+    `
+      export function f(arg: Record<PropertyKey, Promise<string>>) {
+        return arg;
+      }
+    `,
+
+    `
+      export function f<T extends Record<PropertyKey, Promise<string>>>(arg: T) {
+        return arg;
+      }
+    `,
+
+    `
+      export function f<T extends ReturnType<() => string>>(arg: T) {
+        return arg;
+      }
+    `,
```

Comment:
This test is failing for some reason. Seems like this code resolves to `undefined` for `ReturnType`:
```TS
const symbol = type.aliasSymbol ?? type.getSymbol();
```

Am I doing something wrong?

Reply:
Putting a breakpoint on that line of code this in the debugger, I get:

```plaintext
> services.program.getTypeChecker().typeToString(type)
'string'
```

I'm inferring that the type is the result of evaluating `ReturnType<() => string>`, which reduces to `string`. Which actually makes sense to me!

...I don't have a clear thought in my mind as to how to solve for this though. If we skip reporting when `!symbol`, then test cases like this one fail:

```ts
type Arg = number;

const a = function (a: Arg): void {};

export default a;
```

I wonder if the conclusion we should make is to not use the type checker at all? As in, if a type isn't declared or visible from the current file, it's either...

* ...a built-in one
* ...an error type
* ...something globally augmented in

...and for all of those, I _think_ it's fine to not report?

Reply:
well, that's an interesting take
sounds good to me (at least in theory)

Reply:
Done